proposals/records-wip.md

```diff
@@ -76,3 +76,47 @@ each primary constructor parameter to the corresponding property of the other ty
 ```C#
 override Equals(object o) => Equals(o as T);
 ```
+
+## `with` expression
+
+A `with` expression is a new expression using the following syntax.
+
+```antlr
+with_expression
+    : switch_expression
+    | switch_expression 'with' anonymous_object_initializer
+```
+
+A `with` expression allows for "non-destructive mutation", designed to
+produce a copy of the receiver expression with modifications to properties
+listed in the `anonymous_object_initializer`.
+
+A valid `with` expression has a receiver with a non-void type. The receiver type must contain an accessible instance method called `With` with
+the appropriate parameters and return type. It is an error if there are multiple non-override `With` methods. If there are multiple `With` overrides,
+there must be a non-override `With` method, which is the target method. Otherwise, there must be exactly one `With` method.
+
+On the right hand side of the `with` expression is an `anonymous_object_initializer` with a
+sequence of assignments with a field or property of the receiver on the left-hand side of the
+assignment, and an arbitrary expression on the right-hand side which is implicitly convertible to the type
+of the left-hand side.
+
+Given a target `With` method, the return type must be the type of the receiver expression type, or a base type thereof. For each parameter of
```

Comment:
> or a base type thereof

Did you mean derived type?

Reply:
I don't think so... given the example

```C#
class C { }
class B : C {}
```

if `B` has a `With` method, it could return `C` instead of `B`, in that it would be an override of the `With` in `C`, meaning the base type.

Reply:
Should we then spec that even if the return type is `C`, you must return at least `B`, if not a further derived class? Since we have to be able to assign back to a field of `B`? Sure, it's trivially possible to define a `With` method that violates such constraints, but those programs are not in compliance with the spec.

Reply:
This seems fine for now. We can always strengthen it later